### PR TITLE
Add bundled support for styles in template literals

### PR DESCRIPTION
**Pull request by gucong3000**
> Which issue, if any, is this issue related to?

#3264

> Is there anything in the PR that needs further explanation?

e.g. "No, it's self explanatory."

### CODE REVIEW

**Comment by ntwb** [comment]
@evilebottnawi Please read the discussion in #3264 that you yourself had given a 👍 to.

If you have a question on a PR also please add that as a question in a comment or make your PR review a comment, but please don't make these types of questions/comments as "Request Changes" as this is not the intended purpose of the "Request Changes" feature.

**Comment by evilebottnawi** [comment]
@ntwb i support `css-in-js` out of box, but how it is related to `markdown`? We should support `markdown` and `css-in-js` support out of box :confused: And i use `Request Changes`, because `css-in-js` not related to `markdown`.

File: lib/utils/isStandardSyntaxSelector.js [hunk 1]
```diff
@@ -12,7 +12,7 @@ module.exports = function(selector /*: string*/) /*: boolean*/ {
   }
 
   // SCSS placeholder selectors
-  if (selector.indexOf("%") === 0) {
+  if (selector[0] === "%") {
```
Comment (evilebottnawi) [inline lib/utils/isStandardSyntaxSelector.js]: Regression, i can use `.my-class%foo { }`

File: package.json [hunk 2]
```diff
@@ -62,17 +62,17 @@
...
-    "postcss-markdown": "^0.28.0",
     "postcss-media-query-parser": "^0.2.3",
     "postcss-reporter": "^5.0.0",
     "postcss-resolve-nested-selector": "^0.1.1",
     "postcss-safe-parser": "^3.0.1",
     "postcss-sass": "^0.3.0",
     "postcss-scss": "^1.0.2",
     "postcss-selector-parser": "^3.1.0",
-    "postcss-syntax": "^0.28.0",
+    "postcss-styled": "github:gucong3000/postcss-styled#literal_part",
```
Comment (evilebottnawi) [inline package.json]: I think better use stable release for this

**Comment by evilebottnawi** [comment]
Also PR title is misleading :confused: because in package.json we still have `postcss-markdown`

**Comment by ntwb** [comment]
Sorry @evilebottnawi, thanks for that clarification, I missed that myself 😞

File: lib/utils/isStandardSyntaxSelector.js (hunk 1 above)
Comment (gucong3000) [reply lib/utils/isStandardSyntaxSelector.js]:
```js
".my-class%foo".indexOf("%") === 0
// false
```

Comment (Arcanemagus) [reply lib/utils/isStandardSyntaxSelector.js]: `selector.startsWith("%")` is _much_ clearer in the intent... assuming that I'm reading the intent of that code correctly in the first place!

Comment (gucong3000) [reply lib/utils/isStandardSyntaxSelector.js]: fixed in https://github.com/stylelint/stylelint/pull/3405/commits/48d4788f3a6559fb4188b1d1efb4824ced319afe

**Comment by gucong3000** [comment]
@evilebottnawi 
All fixed.

**Comment by jeddy3** [comment]
@gucong3000 Am I right in thinking we also need to bundle `postcss-jsx` if we wish to support the following out-of-the-box?:

```js
import React from "react";

const Component = () => (
  <div style={{top: 0}}></div>
);
```

**Comment by gucong3000** [comment]
> Am I right in thinking we also need to bundle `postcss-jsx` if we wish to support the following out-of-the-box?

Yes, right.

**Comment by jeddy3** [comment]
> Yes, right.

Thanks. Shall we add it in this PR?

**Comment by gucong3000** [comment]
> Shall we add it in this PR?

`postcss-jsx` currently dependencies some non stable version of `@babel/*`.
Maybe we need to wait for the official release of `babel` and then add `postcss-jsx`.

**Comment by jeddy3** [comment]
> Maybe we need to wait for the official release of babel@7 and then add postcss-jsx.

Any ideas when babel@7 is due out?

We could always include `postcss-jsx` and flag it as _experimental_ (which is what we did with the other syntaxes like Less and Sass).

**Comment by gucong3000** [comment]
> We could always include postcss-jsx and flag it as experimental.

If there is no objection, I'd like to add `postcss-jsx` in the next PR

File: lib/rules/declaration-block-no-duplicate-properties/__tests__/index.js [hunk 3]
```diff
@@ -114,6 +114,10 @@ testRule(rule, {
     {
       code: "a { -WEBKIT-border-radius: 12px; -webkit-BORDER-radius: 10px; }",
       message: messages.rejected("-webkit-BORDER-radius")
+    },
```
Comment (jeddy3) [inline lib/rules/declaration-block-no-duplicate-properties/__tests__/index.js]: Can we please pull this out into its own `testRule` block with `syntax: styled`, please?

We should place it at the end of the file, after the `syntax: html` `testRule` block.

File: lib/rules/declaration-block-no-redundant-longhand-properties/__tests__/index.js [hunk 4]
```diff
@@ -79,6 +79,10 @@ testRule(rule, {
       code:
         "a { border-top-width: 1px; border-bottom-width: 1px; border-left-width: 1px; border-right-width: 1px; }",
       message: messages.expected("border-width")
+    },
+    {
```
Comment (jeddy3) [inline lib/rules/declaration-block-no-redundant-longhand-properties/__tests__/index.js]: Can we please pull this out into its own `testRule` block with `syntax: styled`, please?

We should place it at the end of the file.

File: lib/rules/declaration-block-no-shorthand-property-overrides/__tests__/index.js [hunk 5]
```diff
@@ -100,6 +100,10 @@ testRule(rule, {
         "-webKIT-transition",
         "-WEBKIT-transition-property"
       )
+    },
```
Comment (jeddy3) [inline lib/rules/declaration-block-no-shorthand-property-overrides/__tests__/index.js]: Can we please pull this out into its own `testRule` block with `syntax: styled`, please?

We should place it at the end of the file.

File: lib/rules/function-url-no-scheme-relative/__tests__/index.js [hunk 6]
```diff
@@ -69,6 +69,10 @@ testRule(rule, {
     {
       code: "a { background: url(var(--image)); }",
       description: "ignore variable"
+    },
```
Comment (jeddy3) [inline lib/rules/function-url-no-scheme-relative/__tests__/index.js]: Can we please pull this out into its own `testRule` block with `syntax: styled`, please?

We should place it at the end of the file.

File: lib/rules/function-url-quotes/__tests__/index.js [hunk 7]
```diff
@@ -185,6 +185,9 @@ testRule(rule, {
     },
     {
       code: 'a { background: url("") }'
+    },
+    {
```
Comment (jeddy3) [inline lib/rules/function-url-quotes/__tests__/index.js]: Can we please pull this out into its own `testRule` block with `syntax: styled`, please?

We should place it at the end of the file.

File: lib/rules/max-empty-lines/__tests__/index.js [hunk 8]
```diff
@@ -112,6 +112,50 @@ testRule(rule, {
   ]
 });
 
+testRule(rule, {
+  ruleName,
+  config: [1],
```
Comment (jeddy3) [inline lib/rules/max-empty-lines/__tests__/index.js]: Can we add `syntax: styled` here to be explicit, please?

File: lib/rules/max-line-length/__tests__/index.js [hunk 9]
```diff
@@ -169,7 +169,54 @@ a {
...
+      message: messages.expected(20),
+      line: 6,
+      column: 23
+    }
+  ]
+});
+
+testRule(rule, {
+  ruleName,
+  config: [20],
```
Comment (jeddy3) [inline lib/rules/max-line-length/__tests__/index.js]: Can we add `syntax: styled` here please.

File: lib/rules/property-no-unknown/__tests__/index.js [hunk 10]
```diff
@@ -34,6 +34,12 @@ testRule(rule, {
     {
       code: ".foo { *width: 100px; }",
       description: "ignore CSS hacks"
+    },
```
Comment (jeddy3) [inline lib/rules/property-no-unknown/__tests__/index.js]: Can we please pull this out into its own `testRule` block with `syntax: styled`, please?

We should place it at the end of the file.

File: lib/rules/selector-type-no-unknown/__tests__/index.js [hunk 11]
```diff
@@ -132,6 +132,9 @@ testRule(rule, {
     {
       code: ".foo /for/ .bar {}",
       description: "reference combinators"
+    },
```
Comment (jeddy3) [inline lib/rules/selector-type-no-unknown/__tests__/index.js]: Can we please pull this out into its own `testRule` block with `syntax: styled`, please?

We should place it at the end of the file.

**Comment by gucong3000** [comment]
> I've requested some minor changes...

Done.

**Comment by jeddy3** [comment]
@gucong3000 Thank you. Those tests are much easier to scan and consume now.

LGTM.

Has anyone got time to add a 2nd review to this? I'll endeavour to release tomorrow if we get this merged.

**Pull request merged**